pkg/controller/volume/persistentvolume/scheduler_assume_cache.go

```diff
@@ -0,0 +1,283 @@
+	// TODO: index by storageclass
+	c := &assumeCache{store: cache.NewStore(objInfoKeyFunc), description: description}
+
+	// Unit tests don't use informers
+	if informer != nil {
+		informer.AddEventHandler(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc:    c.add,
+				UpdateFunc: c.update,
+				DeleteFunc: c.delete,
+			},
+		)
+	}
+	return c
+}
+
+func (c *assumeCache) add(obj interface{}) {
+	name, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("add: failed to get name: %v", err)
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	objInfo := &objInfo{name: name, latestObj: obj, apiObj: obj}
+	c.store.Update(objInfo)
+}
+
+func (c *assumeCache) update(oldObj interface{}, newObj interface{}) {
+	c.add(newObj)
+}
+
+func (c *assumeCache) delete(obj interface{}) {
+	name, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("delete: failed to get name: %v", err)
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	objInfo := &objInfo{name: name}
+	err = c.store.Delete(objInfo)
+	if err != nil {
+		glog.Errorf("delete: failed to delete %v %v: %v", c.description, name, err)
+	}
+}
+
+func (c *assumeCache) getObjVersion(name string, obj interface{}) (int64, error) {
+	objAccessor, err := meta.Accessor(obj)
+	if err != nil {
+		return -1, err
+	}
+
+	objResourceVersion, err := strconv.ParseInt(objAccessor.GetResourceVersion(), 10, 64)
+	if err != nil {
+		return -1, fmt.Errorf("Error parsing ResourceVersion %q for %v %q: %s", objAccessor.GetResourceVersion(), c.description, name, err)
+	}
+	return objResourceVersion, nil
+}
+
+func (c *assumeCache) getObjInfo(name string) *objInfo {
+	obj, ok, err := c.store.GetByKey(name)
+	if err != nil {
+		glog.Errorf("failed to get objInfo for %v %q: %v", c.description, name, err)
+		return nil
+	}
+	if !ok {
+		return nil
+	}
+
+	objInfo, ok := obj.(*objInfo)
+	if !ok {
+		glog.Errorf("object is not of type objInfo for %v %q", c.description, name)
+		return nil
+	}
+	return objInfo
+}
+
+func (c *assumeCache) Get(objName string) interface{} {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if objInfo := c.getObjInfo(objName); objInfo != nil {
+		return objInfo.latestObj
+	}
+	return nil
+}
+
+func (c *assumeCache) List() []interface{} {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	allObjs := []interface{}{}
+	for _, obj := range c.store.List() {
+		objInfo, ok := obj.(*objInfo)
+		if !ok {
+			glog.Errorf("List: object is not of type objInfo for %v", c.description)
+			continue
+		}
+		allObjs = append(allObjs, objInfo.latestObj)
+	}
+	return allObjs
+}
+
+func (c *assumeCache) Assume(obj interface{}) error {
+	name, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return fmt.Errorf("Assume: failed to get name: %v", err)
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	objInfo := c.getObjInfo(name)
+	if objInfo == nil {
+		return fmt.Errorf("Assume: %v %q not found in cache", c.description, name)
```

Comment:
Errors should start with lowercase (applies to whole file). Unless 'Assume' is name of something and it gets logged from other assume errors too.

Reply:
I was referring to the name of the function, but I can change the errors to not reference function names.